x/posts/keeper/posts.go

```diff
@@ -0,0 +1,183 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/desmos-labs/desmos/v3/x/posts/types"
+)
+
+// SetNextPostID sets the new post id for the given subspace to the store
+func (k Keeper) SetNextPostID(ctx sdk.Context, subspaceID uint64, postID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.NextPostIDStoreKey(subspaceID), types.GetPostIDBytes(postID))
+}
+
+// HasNextPostID tells whether the next post id key existing for the given subspace
+func (k Keeper) HasNextPostID(ctx sdk.Context, subspaceID uint64) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(types.NextPostIDStoreKey(subspaceID))
+}
+
+// GetNextPostID gets the highest post id for the given subspace
+func (k Keeper) GetNextPostID(ctx sdk.Context, subspaceID uint64) (postID uint64, err error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.NextPostIDStoreKey(subspaceID))
+	if bz == nil {
+		return 0, sdkerrors.Wrapf(types.ErrInvalidGenesis, "initial post ID hasn't been set for subspace %d", subspaceID)
+	}
+
+	postID = types.GetPostIDFromBytes(bz)
+	return postID, nil
+}
+
+// DeleteNextPostID removes the post id key for the given subspace
+func (k Keeper) DeleteNextPostID(ctx sdk.Context, subspaceID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.NextPostIDStoreKey(subspaceID))
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidatePostReference checks the post reference to make sure that the referenced
+// post's author has not blocked the user referencing the post
+func (k Keeper) ValidatePostReference(ctx sdk.Context, postAuthor string, subspaceID uint64, referenceID uint64) error {
+	// Make sure the referenced post exists
+	referencedPost, found := k.GetPost(ctx, subspaceID, referenceID)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "post with id %d does not exist", referenceID)
+	}
+
+	// Make sure the original author has not blocked the post author
+	if k.HasUserBlocked(ctx, referencedPost.Author, postAuthor, subspaceID) {
```

Comment:
Should it check the block in genral subspace in general subspace 0?

Reply:
No, it shouldn't. We have designed blocks to be subspace-specific: if you block someone on the subspace with id `0`, then it means that user will not be able to send you DTag transfer requests. But they will be still be able to tag you/reply you inside other subspaces. You should block them there as well if you don't want them to interact with you at all